File: algos/sorts/mergeSort.js
function mergeSort(array){
    var leftIndex = 0;
    var rightIndex = 0;

    //base case
    if(array.length < 2){
        return array
    }

    var left = array.slice(0, array.length / 2);
    var right = array.slice(array.length / 2, array.length);

    left = mergeSort(left);
    right = mergeSort(right);

    return merge(left, right, leftIndex, rightIndex);
}

function merge(leftArray, rightArray, leftIndex, rightIndex){
    var results = [];

    while(leftArray.length > leftIndex && rightArray.length > rightIndex) {
        if(leftArray[leftIndex] > rightArray[rightIndex]){
            results.push(rightArray[rightIndex]);
            rightIndex++;
        } else {
            results.push(leftArray[leftIndex]);
            leftIndex++;
        }
    }

    while(leftArray.length > leftIndex){
        results.push(leftArray[leftIndex]);
        leftIndex++;
    }

    while(rightArray.length > rightIndex){
        results.push(rightArray[rightIndex]);
        rightIndex++;
    }

    return results;
}

function createRandomArray(range, size) {
    var arr = [];
    for(let i = 0; i < size; i++){
        arr[i] = Math.floor(Math.random() * range);
        arr.push(arr[i]);
    }

    return arr
}


console.log(mergeSort([4,3,8,2,5]));
console.log(mergeSort(createRandomArray(500,10)));
console.log(mergeSort(createRandomArray(500,10)));
console.log(mergeSort(createRandomArray(500,10)));

File: algos/sorts/merge_sort.py
def merge_sort(arr):
    left_index = 0
    right_index = 0

    if len(arr) < 2:
        return arr

    left = arr[0:(len(arr)//2)]
    right = arr[len(arr)//2:len(arr)]

    left = merge_sort(left)
    right = merge_sort(right)

    return merge(left, right, left_index, right_index)

def merge(left, right, left_index, right_index):
    results = []

    while len(left) < left_index and len(right) > right_index:
        if left[left_index] < right[right_index]:
            results.append(right[right_index])
            right_index += 1
        else:
            results.append(left[left_index])
            left_index += 1

    while len(left) > left_index:
        results.append(left[left_index])
        left_index += 1

    while len(right) > right_index:
        results.append(right[right_index])
        right_index += 1

    return results